File: Python/Thread/5.9.py
import threading
import time
import random

octet = None
ready = False
mutex = threading.Lock()
cond = threading.Condition(mutex)

def emetteur():
    global octet, ready
    for _ in range(10):
        time.sleep(random.randint(1,3))
        with cond:
            while ready:
                cond.wait()
            octet = random.randint(0,255)
            ready = True
            print(f"Émetteur : octet {octet} envoyé")
            cond.notify()

def recepteur():
    global octet, ready
    for _ in range(10):
        with cond:
            while not ready:
                cond.wait()
            print(f"Récepteur : octet {octet} reçu")
            ready = False
            cond.notify()

t1 = threading.Thread(target=emetteur)
t2 = threading.Thread(target=recepteur)

t1.start()
t2.start()
t1.join()
t2.join()

File: C/Thread/5.9.c
#include <stdio.h>
#include <pthread.h>
#include <unistd.h>
#include <stdlib.h>
#include <time.h>

int octet;
int ready = 0; // indique si l'octet est prêt

pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t cond = PTHREAD_COND_INITIALIZER;

void* emetteur(void* arg)
{
    for(int i=0;i<10;i++){
        sleep(rand()%3 + 1); // temps aléatoire 1-3 sec
        pthread_mutex_lock(&mutex);
        while(ready) 
            pthread_cond_wait(&cond,&mutex); // attend que octet lu
        octet = rand()%256;
        ready = 1;
        printf("Émetteur : octet %d envoyé\n", octet);
        pthread_cond_signal(&cond); // réveille le recepteur
        pthread_mutex_unlock(&mutex);
    }
    return NULL;
}

void* recepteur(void* arg)
{
    for(int i=0;i<10;i++){
        pthread_mutex_lock(&mutex);
        while(!ready) 
            pthread_cond_wait(&cond,&mutex); // attend octet
        printf("Récepteur : octet %d reçu\n", octet);
        ready = 0;
        pthread_cond_signal(&cond); // réveille l'emetteur
        pthread_mutex_unlock(&mutex);
    }
    return NULL;
}

int main(){
    srand(time(NULL));
    pthread_t t1,t2;

    pthread_create(&t1,NULL,emetteur,NULL);
    pthread_create(&t2,NULL,recepteur,NULL);

    pthread_join(t1,NULL);
    pthread_join(t2,NULL);

    return 0;
}